### ISSUE TITLE
BUG: antialiasing produces artefacts with uniform data

### ISSUE THREAD

**Issue by neutrinoceros** [yt-project/yt#3895]
### Bug report

**Bug summary**

By default, plotting a field with a single uniform value often produce noisy patterns. I assume is because the buffer image isn't _exactly_ uniform unless it aligns with the data grid, so some pixels are interpolated and subject to floating point errors.
Furthermore, since we use the buffer's extrema (rather than the data's) as the colorbar min and max, this floating point noise, however insignificant, ends up dominating the visualisation.

Buffer are always 800x800 pixels by default, so it's easy to get misaligned data.
This often happens when plotting initial conditions where a field is still uniform.

**Code for reproduction**

Here's a minimal example with both an aligned buffer and a misaligned one.
```python
import yt
import numpy as np

shape = (16, 16, 1)
data = {"density": np.ones(shape)}
ds = yt.load_uniform_grid(data, shape, bbox=np.array([[0, 1], [0, 1], [0, 1]]))
p = yt.SlicePlot(ds, "z", "density")
for s in (32, 30):
    p.set_buff_size(s)
    p.annotate_title(f"buffer size = {s}")
    p.save(f"/tmp/test_buff_{s}.png")
```

**Actual outcome**

![test_buff_32](https://user-images.githubusercontent.com/14075922/163265741-b045d730-9b6b-44ef-a9d9-a1788c8cbea7.png)
![test_buff_30](https://user-images.githubusercontent.com/14075922/163265756-22c79312-a649-4655-b9e8-6e96d791fa51.png)

**Expected outcome**

So while there _is_ a workaround in this simple case, it's not very easy to discover without a good understanding of yt's internals, and it should probably not be something that happens by default.

I think this could be solved several ways, but whatever the correct solution is, it's worth keeping in mind that we don't want to take away the ability to visualise data that's _actually_ noisy.
An approach I'm considering would be to use the data's extrema rather than the buffer's to control the color range.

**Comment by neutrinoceros** [comment]
> An approach I'm considering would be to use the data's extrema rather than the buffer's to control the color range.

So I made an attempt at this but it had several drawbacks:
- it doesn't integrate properly with zooming (where the data min and max may be very different from the image's)
- it doesn't really work out of the box for off axis projection plots because they internally use `DummyDataSource` objects which contain much more data than what ends up in an image, so it's impractical

A different approach which was suggested to me would be to change the interpolation scheme used in the cartesian pixelizer routine so it doesn't produce such artifacts at all. This seems like the ideal way to resolve this problem. Let's see if it can be done.

**Comment by neutrinoceros** [comment]
So the problematic interpolation lies within the antialising process, and indeed a better solution to the problem I reported originally, from the user side, it to simply turn off antialiasing, using
```python
p.set_antialiasing(False)
```
Though I still think these artefacts shouldn't be present, so I'm keeping this open for now.

**Comment by neutrinoceros** [comment]
So I thought of automatically turning off the "antialiasing" when the selected data is uniform (this is the approach I took in #3902). However this locks the user in a situation where aa can't be turned on again and hurts performance.

Another possible approach would be to enable making single precision image buffers (they would still be computed as double precision, _then_ truncated). This would eliminate the sensitivity to roundoff errors without hurting performance. However it would probably have to be opt-in to maintain backward compatibility.

**Comment by neutrinoceros** [comment]
Alright so after a lot of thinking I've convinced myself that none of the possible fixes is viable. Here's a summary of my reasoning.

First of, let's reject my initial and naive proposed solution (revamping the aa algorithm so to avoid this noise):
aa is performed by looping over data points, not pixels, so we can't really compare weighted values before weighting them, making roundoff errors unavoidable.

Now, let's take a step back and consider API design and what "sensible defaults" mean; 
I think default parametrisation should favour common use cases and require as little change as possible in edge (yet valid) cases. I can see two relevant edge cases here:
- plotting a grid with uniform values (which is what I started this issue with)
- trying to visualise machine-precision level noise (e.g. plotting the divergence of a magnetic field after some time in an MHD simulation)

Currently, in both these cases, users need to learn about antialiasing, since it's on by default and produces noise at the level of the expected signal. So it's one switch they have to learn, and conveniently, it happens to be the same one, so there's reusability and operating the software at an "advanced" level requires a minimal amount of extra knowledge.

Now, the problem is caused by the combination of the following default parameters:
- aa is on by default
- image buffers are in double precision (same level at which aa is performed)

Turning off aa by default helps in both our edge casing, but is hurtful in the general case, so it's a no brainer.
On the other hand, making image buffers single precision by default _doesn't_ hurt the common use case, *and* is a improvement in the first edge case (field with uniform values), making defaults suited for it too.
However, it makes the second edge case more of a pain to learn how to work around: one would need to learn how to turn off aa but also how to set buffer precision back to double.
I don't think it's reasonable to make one edge case easier to work with if it comes at the cost of making another one more difficult.

In other words, I now think that defaults were particularly well chosen here.
